libs/execution/src/lib/blocks/composite-block-executor.ts

```diff
@@ -0,0 +1,223 @@
+// SPDX-FileCopyrightText: 2023 Friedrich-Alexander-Universitat Erlangen-Nurnberg
+//
+// SPDX-License-Identifier: AGPL-3.0-only
+
+import { strict as assert } from 'assert/strict';
+
+import {
+  BlockDefinition,
+  BlocktypePipeline,
+  BlocktypeProperty,
+  CompositeBlocktypeDefinition,
+  EvaluationContext,
+  IOType,
+  InternalValueRepresentation,
+  Valuetype,
+  createValuetype,
+  evaluateExpression,
+  evaluatePropertyValue,
+  getBlocksInTopologicalSorting,
+  getIOType,
+  isCompositeBlocktypeDefinition,
+} from '@jvalue/jayvee-language-server';
+
+import { ExecutionContext } from '../execution-context';
+import { IOTypeImplementation, NONE } from '../types';
+
+// eslint-disable-next-line import/no-cycle
+import { executeBlocks } from './block-execution-util';
+import { AbstractBlockExecutor, BlockExecutor } from './block-executor';
+import { BlockExecutorClass } from './block-executor-class';
+import * as R from './execution-result';
+
+export function createCompositeBlockExecutor(
+  inputType: IOType,
+  outputType: IOType,
+  block: BlockDefinition,
+): BlockExecutorClass<BlockExecutor<IOType, IOType>> {
+  assert(
+    block.type.ref,
+    `Blocktype reference missing for block ${block.name}.`,
+  );
+
+  assert(
+    isCompositeBlocktypeDefinition(block.type.ref),
+    `Blocktype is not a composite block for block ${block.name}.`,
+  );
+
+  const blockTypeReference = block.type.ref;
+
+  return class extends AbstractBlockExecutor<
+    typeof inputType,
+    typeof outputType
+  > {
+    public static readonly type = blockTypeReference.name;
+
+    constructor() {
+      super(inputType, outputType);
+    }
+
+    async doExecute(
+      input: IOTypeImplementation<typeof inputType>,
+      context: ExecutionContext,
+    ): Promise<R.Result<IOTypeImplementation<typeof outputType> | null>> {
+      context.logger.logDebug(
+        `Executing composite block of type ${
+          block.type.ref?.name ?? 'undefined'
+        }`,
+      );
+
+      this.addVariablesToContext(block, blockTypeReference.properties, context);
+
+      const executionOrder = getBlocksInTopologicalSorting(
+        blockTypeReference,
+      ).map((block) => {
+        return { block: block, value: NONE };
+      });
+
+      const executionResult = await executeBlocks(
+        context,
+        executionOrder,
+        input,
+      );
+
+      if (R.isErr(executionResult)) {
+        const diagnosticError = executionResult.left;
+        context.logger.logErrDiagnostic(
+          diagnosticError.message,
+          diagnosticError.diagnostic,
+        );
+      }
+
+      this.removeVariablesFromContext(blockTypeReference.properties, context);
+
+      const pipeline = getPipeline(blockTypeReference);
+
+      if (R.isOk(executionResult)) {
+        // The last block always pipes into the output if it exists
+        const lastBlock = pipeline.blocks.at(-1);
```

Comment:
Can be moved before the if-clause / line 94 can be moved down to have it in a cohesive place.